docs/smart_bits.md

```diff
@@ -0,0 +1,71 @@
+
+It should also be noted that due to the implementation details, expressions on smart types are not actually operable magma types. In fact, the only thing you can do with them is wire them to other smart types, as in `a @= expr`. At the time of wiring, bit-widths and signedness of operands are resolved as a function of the expression itself and the left hand side of the assignment (see below for details). So, a statement like `b @= expr`, means that the generated RTL could different based on the bit-width of `b`.
+
+## Types
+There are two smart type constructors:
+* `SmartBits[bit_width, signed=False]`: returns a SmartBits type with bit-width `bit_width` and signedness `sign`. `sign` defaults to `False`.
+* `SmartBit` which is a literal alias of `SmartBits[1]`.
+
+Expressions built on top of these base smart types are of an intermediate type `SmartExpr`, which can only (a) be used in another expression, such as `dbl = expr + expr`, or (b) be on the right hand side of a wiring assignment. Besides these two operations (and calling `str()` for debugging), these intermediate objects are not meant to be introspected or operated on by the user.
+
+## Operators
+Arithmetic, bit-wise logical, comparison, shift, bit-wise reduction, and concat operators are defined for all smart types and intermediates. All operators are implemented as python-style infix operators (e.g. `a + b`), except for the following:
+* Concat: the function `concat` is a free, variadic function; e.g. `long = concat(x, y, z)`
+* Bit-wise reduce: bit-wise reductions are defined as a member function `reduce`, which takes the reduction operator as an argument; e.g. `short = x.reduce(operator.or_)`
+
+See the end of the document for a table summarizing all operators.
+
+## Inference
+We first define a *partial expression* to be any (potentially compound) expression with smart operands, without wiring, and we define a *complete expression* to be any wiring statement where the left hand side is a base smart type and the right hand side is a partial expression. Every complete expression is dynamically resolved using a scheme to infer the bit-width and signedness of each operand. This is inspired from the [Verilog standard](https://www.eg.bucknell.edu/~csci320/2016-fall/wp-content/uploads/2015/08/verilog-std-1364-2005.pdf) (see Sec. 5.2).
+
+In general, every partial expression appearing in a complete expression is either *self-determined* or *context-determined*. This distinguishes expressions which require the entirety of the complete expression in-order to do inference, or only the partial expression. Each operator may have further specialized inference rules.
```

Comment:
Can a partial expression be used more than once? (e.g. wired to two different endpoints)

Reply:
Yes. Added clarification about this point earlier in the doc.